### PR TITLE
[MIN-2432] Implement Python Enum schema evolution

### PR DESCRIPTION
**Pull request by antongrbin**
# Motivation

Tech doc:
https://docs.google.com/document/d/1p5LUSTWrVBcT9F2wXBgHBDT3OJyOm5BpXAyTpwwDVts/edit#

# Changes

In this PR we actually fix the Python protobuf library not to break when unknown enum value is encountered if ignoreUnknownFields is true.

The design principles for the change were:
* make the change as small as possible, so it's easy to keep up to date with the upstream
* if one element is unknown in an array of elements, other elements should still parse (see unit tests)

Changes:
* Change `_ConvertScalarFieldValue` so it throws a special exception when unknown enum string value is encountered.
* Wrap every invocation of `_ConvertScalarFieldValue` to optionally ignore the new exception thrown.
* Add unit tests

To write this PR, I reached out for input on #python-club [on slack](https://noom.slack.com/archives/C020854DJDU/p1655740069429799)

# Tested

Executed the test plan from here: https://github.com/noom/protobuf/pull/1 to test this within the `protobuf` repo.

Also tested as part of `noom-contracts` repo: https://github.com/noom/noom-contracts/pull/293

### CODE REVIEW

File: python/google/protobuf/internal/json_format_test.py [hunk 1]
```diff
@@ -909,6 +909,26 @@ def testParseEnumValue(self):
...
+  def testParseUnknownEnumStringValueProto2(self):
+    message = json_format_pb2.TestNumbers()
+    text = '{"a": "UNKNOWN_STRING_VALUE"}'
+    json_format.Parse(text, message, ignore_unknown_fields=True)
+    # In proto2 we can see that the field was not set at all.
+    self.assertFalse(message.HasField("a"))
+
+  def testParseUnknownEnumStringValueRepeatedProto3(self):
+    message = json_format_proto3_pb2.TestMessage()
+    text = '{"repeatedEnumValue": ["UNKNOWN_STRING_VALUE", "FOO", "BAR"]}'
```
Comment (ifabijanovic) [inline python/google/protobuf/internal/json_format_test.py:928]: Isn't an array of enums banned?

File: python/google/protobuf/internal/json_format_test.py [hunk 2]
```diff
@@ -909,6 +909,26 @@ def testParseEnumValue(self):
...
+  def testParseUnknownEnumStringValueProto3(self):
+    message = json_format_proto3_pb2.TestMessage()
+    text = '{"enumValue": "UNKNOWN_STRING_VALUE"}'
+    json_format.Parse(text, message, ignore_unknown_fields=True)
+    # In proto3, there is no difference between the default value and 0.
+    self.assertEqual(message.enum_value, 0)
+
+  def testParseUnknownEnumStringValueProto2(self):
+    message = json_format_pb2.TestNumbers()
+    text = '{"a": "UNKNOWN_STRING_VALUE"}'
```
Comment (ifabijanovic) [inline python/google/protobuf/internal/json_format_test.py:921]: curious why the key is `a` here instead of `enumValue` like in other tests?

File: python/google/protobuf/json_format.py [hunk 3]
```diff
@@ -116,6 +117,28 @@ class ParseError(Error):
   """Thrown in case of parsing error."""
 
 
+class _UnknownEnumStringValueParseError(ParseError):
```
Comment (ifabijanovic) [inline python/google/protobuf/json_format.py:120]: Does this class inherit from `ParseError` (it looks like it does)? So we don't break code that tries to catch `ParseError`

File: python/google/protobuf/internal/json_format_test.py [hunk 4]
```diff
@@ -909,6 +909,26 @@ def testParseEnumValue(self):
         'for enum type protobuf_unittest.TestAllTypes.NestedEnum.',
         json_format.Parse, '{"optionalNestedEnum": 12345}', message)
 
+  def testParseUnknownEnumStringValueProto3(self):
```
Comment (ifabijanovic) [inline python/google/protobuf/internal/json_format_test.py:912]: These tests only test deserialisation. Since we are modifying code that can also *set* values, should we add tests for serialisation as well?

File: python/google/protobuf/internal/json_format_test.py (hunk 2 above)
Comment (antongrbin) [reply python/google/protobuf/internal/json_format_test.py:918]: Good catch, I verified that this is OK and added comment.

```
    # Note that in src/google/protobuf/util/json_format.proto::TestNumbers
    # has a field 'optional MyType a = 1', which is different
    # from the test below that are using
    # src/google/protobuf/util/json_format_proto3.proto::TestMessage.
```

File: python/google/protobuf/internal/json_format_test.py (hunk 1 above)
Comment (antongrbin) [reply python/google/protobuf/internal/json_format_test.py:943]: Good question :)
We banned the array of enums in noom-contracts to avoid people getting confused with this behavior. The array length is 2 although there were 3 fields (one of them unknown).

In this repository (protobuf) I am still testing that the behavior that we have with repeated enums is consistent with the behavior in the binary serialization (regardless of the fact that the behavior is confusing).

I'll add the comment in the code.

File: python/google/protobuf/internal/json_format_test.py (hunk 4 above)
Comment (antongrbin) [reply python/google/protobuf/internal/json_format_test.py]: I am unaware that I am changing the code that sets values.

The code changes are in the following functions:
```
_ConvertScalarFieldValue
_ConvertStructMessage
_ConvertWrapperMessage
_ConvertMapFieldValue
```

These functions are called from the following functions transitively:
_ConvertFieldValuePair
Parser::ConvertMessage
_ConvertAnyMessage
ParseDict

This is only parsing of JSON, I never change serializing to JSON or setting the values.
Did I miss something?

File: python/google/protobuf/json_format.py (hunk 3 above)
Comment (antongrbin) [reply python/google/protobuf/json_format.py:120]: Correct. Added a comment about that.

Although I think the implementation will never leak `_UnknownEnumStringValueParseError`, it still gets renamed to `ParseError` in `_ConvertFieldValuePair`.. but I wouldn't rely on that, the code is very hard to follow :) It's safer if I just inherit and then I know I didn't change any public interface

**Review by antongrbin — COMMENTED** [review]
Thank you for the review!
